Added SendGrid output binding for sending emails

benc-uk opened the pull request:
# Description

Added an output binding for SendGrid
This sends simple HTML emails using the SendGrid API. 

Example component
```
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: sendgrid
spec:
  type: bindings.twilio.sendgrid
  metadata:
  - name: emailFrom
    value: "testapp@dapr.io" # optional 
  - name: emailTo
    value: "dave@dapr.io" # optional 
  - name: subject
    value: "Hello!" # optional 
  - name: apiKey
    value: "CHANGEME" # required, this is your SendGrid key
```

The data field in the binding request is used as email body
You can specify any of the metadata properties on the request too (e.g. `emailFrom`, `emailTo`, `subject`)

Example payload
```
{
  "metadata": {
    "emailTo": "changeme@example.net",
    "subject": "An email from Dapr SendGrid binding"
  }, 
  "data": "<h1>Testing Dapr Bindings</h1>This is a test.<br>Bye!"
}
```
Test / demo app is here https://github.com/benc-uk/dapr-sendgrid-test

TODO: Add docs :)

## Issue reference

Fixes https://github.com/dapr/components-contrib/issues/314

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_  PR: https://github.com/dapr/docs/pull/535

yaron2 commented:
@jjcollinge do you mind reviewing this?

jjcollinge commented:
@yaron2 yh np - will do tomorrow

jjcollinge commented on bindings/sendgrid/sendgrid.go:
```diff
@@ -0,0 +1,131 @@
+	if len(sg.metadata.EmailFrom) > 0 {
+		from = mail.NewEmail("", sg.metadata.EmailFrom)
+	}
+	if len(req.Metadata["emailFrom"]) > 0 {
```
this will overwrite the metadata.EmailFrom if set - is that the intended behaviour i.e. code then config? Feels backwards to me

jjcollinge commented on bindings/sendgrid/sendgrid.go:
```diff
@@ -0,0 +1,131 @@
+	if len(sg.metadata.EmailTo) > 0 {
+		to = mail.NewEmail("", sg.metadata.EmailFrom)
+	}
+	if len(req.Metadata["emailTo"]) > 0 {
```
same as above

jjcollinge commented on bindings/sendgrid/sendgrid.go:
```diff
@@ -0,0 +1,131 @@
+	if len(sg.metadata.EmailTo) > 0 {
+		subject = sg.metadata.Subject
+	}
+	if len(req.Metadata["subject"]) > 0 {
```
same as above

jjcollinge commented on bindings/sendgrid/sendgrid.go:
```diff
@@ -0,0 +1,131 @@
+	client := sendgrid.NewSendClient(sg.metadata.APIKey)
+	resp, err := client.Send(message)
+	if err != nil {
+		return err
```
might want to augment this error. At the moment it'll just print the following
```
== APP == ### Response from dapr: 500 Internal Server Error
```
Something like return `fmt.Errorf("error sending email to SendGrid: %+v", err)`

benc-uk replied:
Fixed

jjcollinge commented on bindings/sendgrid/sendgrid.go:
```diff
@@ -0,0 +1,131 @@
+	// Optional properties
+	sgMeta.EmailFrom = meta.Properties["emailFrom"]
+	sgMeta.EmailTo = meta.Properties["emailTo"]
+	sgMeta.EmailFrom = meta.Properties["emailFrom"]
```
this a dupe of line 49?

benc-uk replied:
Yeah dunno what happened there, copy/paste snafu

jjcollinge commented on bindings/sendgrid/sendgrid.go:
```diff
@@ -0,0 +1,131 @@
+	// Check SendGrid response is OK
+	if !(resp.StatusCode >= 200 && resp.StatusCode < 300) {
+		sg.logger.Errorf("status code from SendGrid not ok %d\n", resp.StatusCode)
+		return fmt.Errorf("error from SendGrid: %d", resp.StatusCode)
```
Do we need both of these?

benc-uk replied:
Probably not, just I find the level of detail I get from errors is really limited. 
When you return an error from a binding, I don't know where it goes? I don't see it logged anywhere, and the caller just gets a 500

benc-uk replied:
Ahh it's in the body of the response, shame daprd doesn't log those messages by default
No matter, that's a wider issue.

jjcollinge replied:
Ah  I presumed the runtime logged the returned error as well as returns it to the user - apparently not. Definietly need to improve logging management around components across the board. cc: @youngbupark

benc-uk replied:
Yeah, it's a pain to have to read the body just to see the error message

jjcollinge commented on bindings/sendgrid/sendgrid.go:
```diff
@@ -0,0 +1,131 @@
+	// Optional properties
+	sgMeta.EmailFrom = meta.Properties["emailFrom"]
+	sgMeta.EmailTo = meta.Properties["emailTo"]
+	sgMeta.EmailFrom = meta.Properties["emailFrom"]
```
Is this a dupe of 49?

jjcollinge commented on bindings/sendgrid/sendgrid.go:
```diff
@@ -0,0 +1,131 @@
+	// Check SendGrid response is OK
+	if !(resp.StatusCode >= 200 && resp.StatusCode < 300) {
+		sg.logger.Errorf("status code from SendGrid not ok %d\n", resp.StatusCode)
+		return fmt.Errorf("error from SendGrid: %d", resp.StatusCode)
```
Are both these needed?

jjcollinge commented on bindings/sendgrid/sendgrid.go:
```diff
@@ -0,0 +1,131 @@
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+package sendgrid
```
Do we think sendgrid should be under the twilio namespace like twilio.sms?

benc-uk replied:
I considered this, it was only when I was done I realised Twilio now owns SendGrid 
AFAIK, it doesn't affect the package name, just the registration in `daprd/main.go`

jjcollinge replied:
Yeh I didn't know where else to put it :P - but to be consistent we'd want to move the directory structure to `twilio/sms` and `twilio/sendgrid` - this would change the import path of both in dapr i.e. `"github.com/dapr/components-contrib/bindings/twilio"` would become `"github.com/dapr/components-contrib/bindings/twilio/sms"` at https://github.com/dapr/dapr/blob/76b06ea574d071f95e879b77460ad5c2d697b893/cmd/daprd/main.go#L93.

benc-uk replied:
Don't mind either way.
I did find it slightly confusing as a first time contributor that the names for registered components was sometimes different from the package name & dir structure. But a quick scan of main.go and all became clear

amanbha replied:
I agree, it should go under `twillio/sendgrid`

benc-uk replied:
Done :)

jjcollinge commented on bindings/sendgrid/sendgrid.go:
```diff
@@ -0,0 +1,131 @@
+	// Build email to
+	var to *mail.Email
+	if len(sg.metadata.EmailTo) > 0 {
+		to = mail.NewEmail("", sg.metadata.EmailFrom)
```
This should be EmailTo

benc-uk replied:
Christ this was sloppy, I was copy and paste happy :(

jjcollinge commented on bindings/sendgrid/sendgrid.go:
```diff
@@ -0,0 +1,131 @@
+
+	// Build email subject
+	subject := ""
+	if len(sg.metadata.EmailTo) > 0 {
```
This should be subject

benc-uk replied:
Snap :(

jjcollinge commented:
LGTM - just need to decide on whether this lives under `/bindings/twilio/sendgrid/sendgrid.go` or just `/bindings/sendgrid/sendgrid.go` - I'll leave that to @yaron2

benc-uk commented:
Moved the sms and sendgrid bindings into a twilio folder 
Changed the component registration to `bindings.twilio.sendgrid`

The registration in daprd `main.go` now looks like 
```go
bindings_loader.NewOutput("twilio.sms", func() bindings.OutputBinding {
  return sms.NewSMS(logContrib)
}),
bindings_loader.NewOutput("twilio.sendgrid", func() bindings.OutputBinding {
  return sendgrid.NewSendGrid(logContrib)
})
```
Tested and working

jjcollinge commented:
> 
> 
> Moved the sms and sendgrid bindings into a twilio folder
> Changed the component registration to `bindings.twilio.sendgrid`
> 
> The registration in daprd `main.go` now looks like
> 
> ```go
> import(
> 	"github.com/dapr/components-contrib/bindings/twilio/sms"
> 	"github.com/dapr/components-contrib/bindings/twilio/sendgrid"
> )
> 
> ... ... ...
> 
> bindings_loader.NewOutput("twilio.sms", func() bindings.OutputBinding {
>   return sms.NewSMS(logContrib)
> }),
> bindings_loader.NewOutput("twilio.sendgrid", func() bindings.OutputBinding {
>   return sendgrid.NewSendGrid(logContrib)
> })
> ```
> 
> Tested and working

Cool, thanks

amanbha commented on bindings/twilio/sendgrid/sendgrid.go:
```diff
@@ -0,0 +1,129 @@
+	if len(req.Metadata["emailTo"]) > 0 {
+		to = mail.NewEmail("", req.Metadata["emailTo"])
+	}
+	if to == nil {
```
Could you add cc and bcc support as well for completeness.

benc-uk replied:
Added now :)

amanbha commented on bindings/twilio/sendgrid/sendgrid.go:
```diff
@@ -0,0 +1,129 @@
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+package sendgrid
```
Please open a doc issue for this as well.

benc-uk replied:
Just checking do you want me to set up a PR with some extra docs, or just an issue to discuss it?

amanbha replied:
Doc issue to start with. PR would be great!

benc-uk replied:
@amanbha Hi, I've created a PR for the docs. Let me know if you need an issue to track it too https://github.com/dapr/docs/pull/535

amanbha replied:
Great @benc-uk Thanks. I am merging it, please make the corresponding changes in dapr runtime as well.

yaron2 commented:
Is sendgrid.go still under /sendgrid? It should move under /twilio/sendgrid.go

yaron2 commented:
And /twilio/sms.go

benc-uk commented:
@yaron2 I moved them around https://github.com/dapr/components-contrib/pull/308#issuecomment-615190383

yaron2 commented:
Ahh.. just to be clear I'm not hallucinating :)

Is it under /twilio/sendgrid.go? Or /twilio/sendgrid/sendgrid.go? It should be the former

benc-uk commented:
`twilio/sms/sms.go`
and
`twilio/sendgrid/sendgrid.go`

This is the convention used by all the other directories that have multiple bindings. e.g. azure, aws, gcp

Otherwise we have two differently named packages in the same directory, which feels wrong

yaron2 commented:
LGTM
